### PR TITLE
added try/except block for NotesSerializerMixin's get_notes_url()

### PR DESCRIPTION
**Pull request by HanlinMiao**
# Closes: #2500 
# What's Changed

# TODO

- [ ] Explanation of Change(s)
- [ ] Added change log fragment(s) (for more information see [the documentation](https://docs.nautobot.com/projects/core/en/stable/development/#creating-changelog-fragments))
- [ ] Attached Screenshots, Payload Example
- [ ] Unit, Integration Tests
- [ ] Documentation Updates (when adding/changing features)
- [ ] Example Plugin Updates (when adding/changing features)
- [ ] Outline Remaining Work, Constraints from Design

### CODE REVIEW

File: nautobot/extras/api/serializers.py [hunk 1]
```diff
@@ -128,8 +129,11 @@ def get_field_names(self, declared_fields, info):
 
     @extend_schema_field(serializers.URLField())
     def get_notes_url(self, instance):
-        notes_url = get_route_for_model(instance, "notes", api=True)
-        return reverse(notes_url, args=[instance.id], request=self.context["request"])
+        try:
+            notes_url = get_route_for_model(instance, "notes", api=True)
+            return reverse(notes_url, args=[instance.id], request=self.context["request"])
+        except NoReverseMatch:
+            return "Notes feature is not available for this model. Please implement it before including NotesSerializerMixin"
```
Comment (glennmatthews) [inline nautobot/extras/api/serializers.py:136]: Seems weird to me to either return a usable URL or else return a human-readable string? If anything, would it be better to *log* the human-readable message and return None?

File: nautobot/extras/tests/test_plugins.py [hunk 2]
```diff
@@ -1,11 +1,11 @@
-from unittest import skipIf
+from unittest import mock, skipIf
 
 from django.conf import settings
 from django.contrib.contenttypes.models import ContentType
 from django.core.exceptions import ValidationError
 from django.template import engines
 from django.test import override_settings
-from django.urls import reverse
+from django.urls import reverse, NoReverseMatch
```
Comment (timizuoebideri1) [inline nautobot/extras/tests/test_plugins.py:8]:
```suggestion
from django.urls import NoReverseMatch, reverse
```

File: nautobot/extras/api/serializers.py [hunk 3]
```diff
@@ -113,6 +116,8 @@
 # Mixins and Base Classes
 #
 
+logger = logging.getLogger("nautobot.extras.api")
```
Comment (glennmatthews) [inline nautobot/extras/api/serializers.py:119]:
```suggestion
logger = logging.getLogger(__name__)
```

File: nautobot/extras/tests/test_plugins.py [hunk 4]
```diff
@@ -439,6 +440,37 @@ def test_api_urls(self):
...
+        obj_perm.save()
+        obj_perm.users.add(self.user)
+        obj_perm.object_types.add(ContentType.objects.get_for_model(self.model))
+        url = self._get_detail_url(instance)
+
+        # Check if log messages and API data is properly rendered
+        with self.assertLogs(logger="nautobot.extras.api", level="WARNING") as cm:
+            response = self.client.get(url, **self.header)
+            self.assertHttpStatus(response, 200)
+            self.assertIn('"notes_url":null', response.content.decode(response.charset))
```
Comment (glennmatthews) [inline nautobot/extras/tests/test_plugins.py:466]: Having a hard time following this one. What's it looking for? Can we provide a more verbose match string?

File: nautobot/extras/tests/test_plugins.py [hunk 5]
```diff
@@ -439,6 +440,37 @@ def test_api_urls(self):
...
+            self.assertIn("notes_url", response.data)
+            self.assertIn("None", str(response.data["notes_url"]))
```
Comment (glennmatthews) [inline nautobot/extras/tests/test_plugins.py:468]: Can we use `assertIsNone` instead?

File: nautobot/extras/api/serializers.py [hunk 6]
```diff
@@ -128,8 +133,14 @@ def get_field_names(self, declared_fields, info):
...
     @extend_schema_field(serializers.URLField())
     def get_notes_url(self, instance):
-        notes_url = get_route_for_model(instance, "notes", api=True)
-        return reverse(notes_url, args=[instance.id], request=self.context["request"])
+        try:
+            notes_url = get_route_for_model(instance, "notes", api=True)
+            return reverse(notes_url, args=[instance.id], request=self.context["request"])
+        except NoReverseMatch:
+            logger.warning(
+                "Notes feature is not available for this model. Please implement it before including NotesSerializerMixin"
```
Comment (glennmatthews) [inline nautobot/extras/api/serializers.py:141]: Would be good to make this more detailed - include the name of the model in question, and explain what "implement it" entails?

File: nautobot/utilities/templatetags/buttons.py [hunk 7]
```diff
@@ -1,5 +1,5 @@
 from django import template
-from django.urls import reverse, NoReverseMatch
+from django.urls import NoReverseMatch, reverse
```
Comment (glennmatthews) [inline nautobot/utilities/templatetags/buttons.py:2]: Please revert unrelated change.

File: nautobot/utilities/testing/views.py [hunk 8]
```diff
@@ -7,7 +7,7 @@
 from django.core.exceptions import ObjectDoesNotExist
 from django.core.files.uploadedfile import SimpleUploadedFile
 from django.test import TestCase as _TestCase, override_settings, tag
-from django.urls import reverse, NoReverseMatch
+from django.urls import NoReverseMatch, reverse
```
Comment (glennmatthews) [inline nautobot/utilities/testing/views.py:10]: Please revert unrelated change.

File: nautobot/extras/tests/test_plugins.py (hunk 4 above)
Comment (HanlinMiao) [reply nautobot/extras/tests/test_plugins.py:466]: This was testing if `return None` is rendered correctly in the browsable API page as `"notes_url": null`. But I dont think its needed since we already do `self.assertIsNone(response.data["notes_url"])`
Screen Shot 2022-11-08 at 10 21 14 PM

File: nautobot/extras/api/serializers.py [hunk 9]
```diff
@@ -128,8 +133,14 @@ def get_field_names(self, declared_fields, info):
...
+            )
+            return None
```
[the remark this answers is not in the document]
Comment (HanlinMiao) [reply nautobot/extras/api/serializers.py:143]: I manually took the notes route out for `CircuitTermination`. This is what the terminal output and browsable API page look like. Im fairly confident it will solve the problem of `get_notes_url()` getting called on without the model having a notes route.
Screen Shot 2022-11-10 at 5 24 05 PM
Screen Shot 2022-11-10 at 5 24 35 PM

**Review by timizuoebideri1 — APPROVED** [review]

**Comment by smk4664** [comment]
NotesSerializerMixin already does a check to ensure that the model has the Notes field. The issue is when the model has the field, but there is no View to route to. Perhaps the message should read to ensure both that the model has the  Notes mixin and the ViewSet has the NotesMixin. Or alternatively point to NautobotModelViewSet.

**Review by glennmatthews — APPROVED** [review]
